### PR TITLE
Revert "[MLIR][Linalg] Re-land linalg.matmul move to ODS. + Remove/update failing obsolete OpDSL tests. (#115319)"

### PR DESCRIPTION
**Pull request by Groverkss**
This reverts commit 3ad0148020ca91cc288bffd8ad36e25f7555a3bb.

As a consequence, c02b8a01b7caf2e4ffe17a123f1bcf59192e4b39 also had to be reverted, which is a follow up fixup to this code.

This patch breaks linalg structured ops. 

- It changes the printer to hide additional attributes, which weren't hidden before: "indexing_maps".
- It changes the build of every linalg structured op to have an indexing map for matmul.

These changes combined, hide the problem until you print the operation in it's generic form.

Reproducer:

```mlir
func.func public @bug(%arg0 : tensor<5x10x20xf32>, %arg1 : tensor<5x20x40xf32>, %arg3 : tensor<5x10x40xf32>) -> tensor<5x10x40xf32> {
  %out = linalg.batch_matmul ins(%arg0, %arg1 : tensor<5x10x20xf32>, tensor<5x20x40xf32>)
      outs(%arg3 : tensor<5x10x40xf32>) -> tensor<5x10x40xf32>
  func.return %out : tensor<5x10x40xf32>
}
```

Prints fine, with `mlir-opt <file>`, but if you do `mlir-opt --mlir-print-op-generic <file>`:

```
#map = affine_map<(d0, d1, d2) -> (d0, d2)>
#map1 = affine_map<(d0, d1, d2) -> (d2, d1)>
#map2 = affine_map<(d0, d1, d2) -> (d0, d1)>
#map3 = affine_map<(d0, d1, d2, d3) -> (d0, d1, d3)>
#map4 = affine_map<(d0, d1, d2, d3) -> (d0, d3, d2)>
#map5 = affine_map<(d0, d1, d2, d3) -> (d0, d1, d2)>
"builtin.module"() ({
  "func.func"() <{function_type = (tensor<5x10x20xf32>, tensor<5x20x40xf32>, tensor<5x10x40xf32>) -> tensor<5x10x40xf32>, sym_name = "bug", sym_visibility = "public"}> ({
  ^bb0(%arg0: tensor<5x10x20xf32>, %arg1: tensor<5x20x40xf32>, %arg2: tensor<5x10x40xf32>):
    %0 = "linalg.batch_matmul"(%arg0, %arg1, %arg2) <{operandSegmentSizes = array<i32: 2, 1>}> ({
    ^bb0(%arg3: f32, %arg4: f32, %arg5: f32):
      %1 = "arith.mulf"(%arg3, %arg4) <{fastmath = #arith.fastmath<none>}> : (f32, f32) -> f32
      %2 = "arith.addf"(%arg5, %1) <{fastmath = #arith.fastmath<none>}> : (f32, f32) -> f32
      "linalg.yield"(%2) : (f32) -> ()
    }) {indexing_maps = [#map, #map1, #map2], linalg.memoized_indexing_maps = [#map3, #map4, #map5]} : (tensor<5x10x20xf32>, tensor<5x20x40xf32>, tensor<5x10x40xf32>) -> tensor<5x10x40xf32>
    "func.return"(%0) : (tensor<5x10x40xf32>) -> ()
  }) : () -> ()
}) : () -> ()
```

The batch_matmul operation's builder now always inserts a indexing_map which is unrelated to the operation itself. This was caught when a transformation from one LinalgStructuredOp to another, tried to pass it's attributes to the other ops builder and there were multiple indexing_map attributes in the result.

### CODE REVIEW

File: mlir/tools/mlir-linalg-ods-gen/mlir-linalg-ods-yaml-gen.cpp [hunk 1]
```diff
@@ -678,11 +678,7 @@ ParseResult {0}::parse(OpAsmParser &parser, OperationState &result) {{
     {0}::getNumRegionArgs(), {0}::getRegionBuilder());
 }
 void {0}::print(OpAsmPrinter &p) {{
-  SmallVector<StringRef, 3> elidedAttrs = {{"operandSegmentSizes",
-                                           "linalg.memoized_indexing_maps",
-                                           "indexing_maps"};
-  ::printNamedStructuredOp(p, getOperation(), getInputs(), getOutputs(),
-                           elidedAttrs);
+  ::printNamedStructuredOp(p, getOperation(), getInputs(), getOutputs());
```
Comment (Groverkss) [inline mlir/tools/mlir-linalg-ods-gen/mlir-linalg-ods-yaml-gen.cpp:681]: The printer should not be eliding an attribute like indexing_maps

File: mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp [hunk 2]
```diff
@@ -195,20 +168,6 @@ static void buildStructuredOp(
...
-    for (mlir::AffineMap map : *indexingMaps) {
-      // Convert each AffineMap to an AffineMapAttr
-      indexingMapsAttrVal.push_back(AffineMapAttr::get(map));
-    }
-    state.addAttribute("indexing_maps", b.getArrayAttr(indexingMapsAttrVal));
-  } else {
-    indexingMapsAttrVal = getDefaultIndexingMapAttr(b.getContext());
-    state.addAttribute("indexing_maps", b.getArrayAttr(indexingMapsAttrVal));
-  }
-
```
Comment (Groverkss) [inline mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:211]: This is only valid for MatmulOp and should not be part of this method.

File: mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp [hunk 3]
```diff
@@ -155,36 +149,15 @@ static void fillStructuredOpRegion(OpBuilder &opBuilder, Region &region,
...
-  indexingMaps.push_back(AffineMap::get(3, 0, {d2, d1}, context));
-  indexingMaps.push_back(AffineMap::get(3, 0, {d0, d1}, context));
-  return indexingMaps;
-}
-
-/// Wrapper to return the typical indexing map array attribute for MatmulOp.
-static SmallVector<Attribute> getDefaultIndexingMapAttr(MLIRContext *context) {
-  return llvm::map_to_vector(
-      getDefaultIndexingMapsForMatmul(context),
-      [](AffineMap map) -> Attribute { return AffineMapAttr::get(map); });
```
Comment (Groverkss) [inline mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:175]: This is a misleading name for a method. It should be getDefaultMatmulIndexingMapAttr. Also, a rewriter can do this for you, so this method is redundant.

**Review by qedawkins — APPROVED** [review]
I think the extraneous indexing maps is the main problem here. Let's also wait a bit for others to take a look because there was a long discussion on the initial PR.

**Review by Groverkss — COMMENTED** [review]
I inlined commented the buggy parts of the patch.

**Comment by Groverkss** [comment]
I'll wait for others to comment. While it doesn't look like a trivial fix, I'm okay with a forward fix as well, but I will let others decide.

**Comment by Groverkss** [comment]
> This patch got reverted before. If we keep reverting, it will be hard to make forward progress. Can you propose fixes to the problems you mention in other PRs instead of reverting?
> 
> They all look reasonable solutions to the problems you mention, so I don't see any friction being created from the follow up patches. I'll be happy to review such patches as soon as possible.

Forward fix: https://github.com/llvm/llvm-project/pull/116412

**Pull request closed without merging**